[SPARK-20213][SQL][UI] Fix DataFrameWriter operations in SQL UI tab.

rdblue opened the pull request:
## What changes were proposed in this pull request?

Wraps `DataFrameWriter` operations in `SQLExecution.withNewExecutionId` so that `SparkListenerSQLExecutionStart` and `SparkListenerSQLExecutionEnd` are sent and the query shows up in the SQL tab of the UI.

## How was this patch tested?

Tested by hand that `insertInto` results in queries in the SQL tab.

rdblue committed:
SPARK-20213: Fix DataFrameWriter operations in SQL UI tab.

Not wrapping the execution started by DataFrameWriter in
SQLExecution.withNewExecutionId causes those executions to not show up
in the SQL UI tab because the SparkListenerSQLExecutionStart and
SparkListenerSQLExecutionEnd messages are never sent.

SparkQA commented:
**[Test build #75549 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75549/testReport)** for PR 17540 at commit [`f9342b5`](https://github.com/apache/spark/commit/f9342b5fea4fd3dbe32c12930f96aa9af38c386b).

srowen commented:
Looks closely related to https://github.com/apache/spark/pull/17535 ?

rdblue commented:
@srowen, agreed. Closely related but not the same code paths. The question is: when should `withNewExecutionId` get called?

I'm running the test suite now and this patch causes test failures when `withNewExecutionId` is called twice; once in `DataFrameWriter` and once in `InsertIntoHadoopFsRelationCommand`. It looks like the call has now been littered about the codebase (e.g. in `InsertIntoHadoopFsRelationCommand` and other execution nodes) to fix this problem on certain operations, so we should decide where it should be used and fix tests around that.

The reason why I added it to `DataFrameWriter` is that it is called in `Dataset` actions, and it makes sense to call it once from where an action is started. I think it makes the most sense for action methods, like `Dataset#collect` or `DataFrameWriter#insertInto` to minimize the number of places we need to add it. I don't think this is a concern that should be addressed by the execution plan.

SparkQA commented:
**[Test build #75549 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75549/testReport)** for PR 17540 at commit [`f9342b5`](https://github.com/apache/spark/commit/f9342b5fea4fd3dbe32c12930f96aa9af38c386b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75549/
Test FAILed.

rdblue commented:
@cloud-fan, can you look at this? What do you think about the question above: when should `withNewExecutionId` get called?

rdblue committed:
SPARK-20213: Add check to ensure exection ID is set.

SparkQA commented:
**[Test build #75579 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75579/testReport)** for PR 17540 at commit [`a3296a2`](https://github.com/apache/spark/commit/a3296a256d72899aee886816e732b8ad633cf170).

SparkQA commented:
**[Test build #75579 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75579/testReport)** for PR 17540 at commit [`a3296a2`](https://github.com/apache/spark/commit/a3296a256d72899aee886816e732b8ad633cf170).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75579/
Test FAILed.

rdblue committed:
SPARK-20213: Add withNewExecutionId to commands run by spark.sql.

SparkQA commented:
**[Test build #75581 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75581/testReport)** for PR 17540 at commit [`429edfb`](https://github.com/apache/spark/commit/429edfb6c9e45e1cf2210629a6d2db2d8e91d72b).

rdblue commented:
The new test failures are caused by a check I inserted. Moving where `withNewExecutionId` gets called could result in missing SQL queries in the UI, so anywhere I'm replacing `withNewExecutionId` with `checkSQLExecutionId` that will cause tests (and only tests) to fail. That way, we can catch all of the call stacks that should have it.

This caught problems in SQL command execution and I've added a patch to fix it.

SparkQA commented:
**[Test build #75581 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75581/testReport)** for PR 17540 at commit [`429edfb`](https://github.com/apache/spark/commit/429edfb6c9e45e1cf2210629a6d2db2d8e91d72b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75581/
Test FAILed.

cloud-fan commented:
The `withNewExecutionId` was added at https://github.com/rxin/spark/commit/1b0317f64cfe99ff70580eeb99753cd0d31f849a#diff-89b9796aae086e790ddd9351f0db8115R134 .

The execution id is used to track all jobs that belong to the same query, so I think it makes sense to call `withExecutionId` at action methods like `Dataset#collect` or `DataFrameWriter#insertInto`

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -180,9 +180,13 @@ class Dataset[T] private[sql](
     // to happen right away to let these side effects take place eagerly.
     queryExecution.analyzed match {
       case c: Command =>
-        LocalRelation(c.output, queryExecution.executedPlan.executeCollect())
```
how about `LocalRelation(c.output, withAction("collect")(_. executeCollect()))`

cloud-fan replied:
Actually do we need to do this? most `Command`s are just local operations(talking with metastore).

rdblue replied:
Yeah, the check I added to ensure we get the same results in the SQL tab has [several hundred failures](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75579/testReport/) that go through this. Looks like the path is almost always `spark.sql` when the SQL statement is a command like CTAS.

I like your version and will update.

cloud-fan commented:
LGTM, @rdblue the failed tests are thrift server tests, which are hard to debug. You can run hive tests locally and see what failed.(usually failed thrift server tests means we have failed hive tests)

rdblue commented:
Thanks for the review! I'll get the thrift-server tests fixed up next week.

rdblue committed:
SPARK-20213: Allow nested withNewExecutionId.

SparkQA commented:
**[Test build #75675 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75675/testReport)** for PR 17540 at commit [`7cb7d4e`](https://github.com/apache/spark/commit/7cb7d4e94d631a98f8790dcc364fc83b1d1bc3b8).

rdblue committed:
SPARK-20213: Use withAction instead of withNewExecutionId.

Other methods in Dataset use withAction to start an execution, which
includes withNewExecutionId and other tracking.

SparkQA commented:
**[Test build #75676 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75676/testReport)** for PR 17540 at commit [`68cc2b3`](https://github.com/apache/spark/commit/68cc2b37918942a037a4e5f93155013b0e805ad6).

SparkQA commented:
**[Test build #75675 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75675/testReport)** for PR 17540 at commit [`7cb7d4e`](https://github.com/apache/spark/commit/7cb7d4e94d631a98f8790dcc364fc83b1d1bc3b8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75675/
Test FAILed.

SparkQA commented:
**[Test build #75676 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75676/testReport)** for PR 17540 at commit [`68cc2b3`](https://github.com/apache/spark/commit/68cc2b37918942a037a4e5f93155013b0e805ad6).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75676/
Test FAILed.

rdblue committed:
SPARK-20213: Throw errors during testing for nested execution.

SparkQA commented:
**[Test build #75677 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75677/testReport)** for PR 17540 at commit [`ce0dbe7`](https://github.com/apache/spark/commit/ce0dbe7c6080dc3cdcee30ac404222ad3480b2ca).

SparkQA commented:
**[Test build #75677 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75677/testReport)** for PR 17540 at commit [`ce0dbe7`](https://github.com/apache/spark/commit/ce0dbe7c6080dc3cdcee30ac404222ad3480b2ca).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75677/
Test FAILed.

rdblue committed:
SPARK-20213: Add withNewExecutionId to test method.

SparkQA commented:
**[Test build #75681 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75681/testReport)** for PR 17540 at commit [`7910825`](https://github.com/apache/spark/commit/7910825ce7e0cc7d8ba1eb9f6e534b087815f294).

SparkQA commented:
**[Test build #75681 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75681/testReport)** for PR 17540 at commit [`7910825`](https://github.com/apache/spark/commit/7910825ce7e0cc7d8ba1eb9f6e534b087815f294).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75681/
Test FAILed.

rdblue committed:
SPARK-20213: Fix failing tests.

rdblue committed:
SPARK-20213: Call withNewExecutionId from StreamExecution.

SparkQA commented:
**[Test build #75748 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75748/testReport)** for PR 17540 at commit [`2cee6b7`](https://github.com/apache/spark/commit/2cee6b7de08fc433180abc4e3f40d5ea3a0a7d41).

SparkQA commented:
**[Test build #75748 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75748/testReport)** for PR 17540 at commit [`2cee6b7`](https://github.com/apache/spark/commit/2cee6b7de08fc433180abc4e3f40d5ea3a0a7d41).
 * This patch **fails Scala style tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75748/
Test FAILed.

SparkQA commented:
**[Test build #75749 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75749/testReport)** for PR 17540 at commit [`4f3a02b`](https://github.com/apache/spark/commit/4f3a02b1005bfd1a581a055c9d3af3aac98b0119).

SparkQA commented:
**[Test build #75749 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75749/testReport)** for PR 17540 at commit [`4f3a02b`](https://github.com/apache/spark/commit/4f3a02b1005bfd1a581a055c9d3af3aac98b0119).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75749/
Test FAILed.

SparkQA commented:
**[Test build #75780 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75780/testReport)** for PR 17540 at commit [`cfe4e2c`](https://github.com/apache/spark/commit/cfe4e2cd3389eb8ab37edce82fc826383e0a8351).

SparkQA commented:
**[Test build #75781 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75781/testReport)** for PR 17540 at commit [`36da73b`](https://github.com/apache/spark/commit/36da73b413d2fe33dc339c6137be6efcb044bb8f).

SparkQA commented:
**[Test build #75780 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75780/testReport)** for PR 17540 at commit [`cfe4e2c`](https://github.com/apache/spark/commit/cfe4e2cd3389eb8ab37edce82fc826383e0a8351).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75780/
Test FAILed.

SparkQA commented:
**[Test build #75781 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75781/testReport)** for PR 17540 at commit [`36da73b`](https://github.com/apache/spark/commit/36da73b413d2fe33dc339c6137be6efcb044bb8f).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75781/
Test FAILed.

rdblue committed:
SPARK-20213: Fix SQLListener test.

Opening up metrics in the last commit caused this test to fail.

SparkQA commented:
**[Test build #75786 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75786/testReport)** for PR 17540 at commit [`a822309`](https://github.com/apache/spark/commit/a82230974d9189e419f6c1ba6d32d97982e28cd1).

SparkQA commented:
**[Test build #75786 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75786/testReport)** for PR 17540 at commit [`a822309`](https://github.com/apache/spark/commit/a82230974d9189e419f6c1ba6d32d97982e28cd1).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75786/
Test FAILed.

rdblue committed:
SPARK-20213: Remove SPARK-10548 repro test.

There is no longer a need for this test because the case it reproduces
is fixed in two ways. First, there is a fix for the original problem
where execution IDs leaked across threads. Second, nested execution IDs
no longer cause exceptions outside of tests because there is no reason
to fail at runtime just to keep the SQL tab clean.

This test was passing in Maven, but failing in SBT.

SparkQA commented:
**[Test build #75813 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75813/testReport)** for PR 17540 at commit [`30fa4fc`](https://github.com/apache/spark/commit/30fa4fc8603e68f9295fc65e573f96140bb04ac6).

SparkQA commented:
**[Test build #75813 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75813/testReport)** for PR 17540 at commit [`30fa4fc`](https://github.com/apache/spark/commit/30fa4fc8603e68f9295fc65e573f96140bb04ac6).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75813/
Test FAILed.

rdblue commented:
@cloud-fan, could you have another look at this?

There are a few new changes:
* withNewExecutionId now warns instead of throwing an exception, but still throws exceptions if spark.testing is defined
* SQLExecution.nested allows nested execution IDs without test failures or warnings. This is needed because several places will nest when withNewExceptionId is called at the high-level operations. CacheTableCommand is an example.

Over the last week, I've fixed nearly all of the tests. The remaining failure, SQLExecutionSuite.concurrent query execution (SPARK-10548), is fixed in maven, but fails in SBT. The problem is that exceptions are now only thrown if `spark.testing` is defined, and for some reason adding it to the test's SparkSession or SparkContext doesn't work on Jenkins. Because this test is reproducing a case that now will never happen for two reasons (the original multi-threading fix and throw only if spark.testing), I'd like to simply remove it. Let me know what you think about that.

Other changes to look at:
* `SQLMetricsSuite.save metrics` started failing because there is a nested execution ID. This is because there are two SQL physical plans. The first, `ExecutedCommandExec` links in a logical plan that is turned into a second physical plan *at runtime*. This means that the inner plan can't report the metrics that will be collected when analyzing the outer plan because it doesn't exist yet. The long-term solution is to fix `ExecutedCommandExec`, but for now this accepts any metrics created by the inner plan.
* `StreamExecution` wasn't calling `withNewExecutionId` and was caught by the new assertion. I added the call around the entire execution so that there isn't a new SQL execution for every batch. This required creating a special `queryExecution` to pass in.
* `DataFrameCallbackSuite` had to be updated to include commands that were previously not registered in the SQL tab. The new SQL executions are for dropping tables, so the result looks more correct than before.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -180,9 +180,9 @@ class Dataset[T] private[sql](
+        LocalRelation(c.output, withAction("collect", queryExecution)(_.executeCollect()))
       case u @ Union(children) if children.forall(_.isInstanceOf[Command]) =>
-        LocalRelation(u.output, queryExecution.executedPlan.executeCollect())
+        LocalRelation(u.output, withAction("collect", queryExecution)(_.executeCollect()))
```
shall we only add execution id for commands that will trigger execution? AFAIK there are 3 commands: `CreateDataSourceTableAsSelectCommand`, `CreateHiveTableAsSelectCommand` and `CacheTable`. We can call `SQLExecution.withNewExecutionId` inside these 3 commands.

Then we don't need to worry about nested execution.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:
```diff
@@ -39,6 +39,32 @@ object SQLExecution {
 
+  private val testing = sys.props.contains("spark.testing")
+
+  private[sql] def checkSQLExecutionId(sparkSession: SparkSession): Unit = {
```
this is only called in `FileFormatWirter`, is there any other places we need to consider?

rdblue replied:
To keep this PR from growing too big, I want to just use it where I've removed `withNewExecutionId` to check for regressions. I'll follow up with another PR with more checks.

cloud-fan commented:
yea let's remove that test

rdblue commented:
Removed the failing SPARK-10548 test and rebased.

SparkQA commented:
**[Test build #75860 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75860/testReport)** for PR 17540 at commit [`901cec8`](https://github.com/apache/spark/commit/901cec88d278ea63072f32dd3236780c6bb2d449).

SparkQA commented:
**[Test build #75860 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75860/testReport)** for PR 17540 at commit [`901cec8`](https://github.com/apache/spark/commit/901cec88d278ea63072f32dd3236780c6bb2d449).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75860/
Test FAILed.

rdblue commented:
@cloud-fan, I've updated this and in Jenkins all tests are passing. Where are the results for the failing PySpark tests?

gatorsmile commented:
retest this please

SparkQA commented:
**[Test build #75865 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75865/testReport)** for PR 17540 at commit [`901cec8`](https://github.com/apache/spark/commit/901cec88d278ea63072f32dd3236780c6bb2d449).

SparkQA commented:
**[Test build #75865 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75865/testReport)** for PR 17540 at commit [`901cec8`](https://github.com/apache/spark/commit/901cec88d278ea63072f32dd3236780c6bb2d449).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75865/
Test FAILed.

cloud-fan commented:
@rdblue You need to click the "consle output" tag on the left to see the pyspark failures. Here it is
```
ERROR: test_stream_exception (pyspark.sql.tests.SQLTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/workspace/SparkPullRequestBuilder/python/pyspark/sql/tests.py", line 1438, in test_stream_exception
    sq.processAllAvailable()
  File "/home/jenkins/workspace/SparkPullRequestBuilder/python/pyspark/sql/streaming.py", line 149, in processAllAvailable
    return self._jsq.processAllAvailable()
  File "/home/jenkins/workspace/SparkPullRequestBuilder/python/lib/py4j-0.10.4-src.zip/py4j/java_gateway.py", line 1133, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/home/jenkins/workspace/SparkPullRequestBuilder/python/pyspark/sql/utils.py", line 75, in deco
    raise StreamingQueryException(s.split(': ', 1)[1], stackTrace)
StreamingQueryException: u'spark.sql.execution.id is already set\n=== Streaming Query ===\nIdentifier: query_explain [id = 568c91fc-0d6d-4fc8-a7d7-db94b6603c8a, runId = e646eee0-22c5-4620-9987-8a0d1c682476]\nCurrent Committed Offsets: {}\nCurrent Available Offsets: {FileStreamSource[file:/home/jenkins/workspace/SparkPullRequestBuilder/python/test_support/sql/streaming]: {"logOffset":0}}\n\nCurrent State: ACTIVE\nThread State: RUNNABLE\n\nLogical Plan:\nFileStreamSource[file:/home/jenkins/workspace/SparkPullRequestBuilder/python/test_support/sql/streaming]'
```

rdblue committed:
SPARK-20213: Fix pyspark test failure.

SparkQA commented:
**[Test build #75996 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75996/testReport)** for PR 17540 at commit [`adbbc7a`](https://github.com/apache/spark/commit/adbbc7a66eabe72af2887015cbf6eb6e8553e2c6).

SparkQA commented:
**[Test build #75996 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75996/testReport)** for PR 17540 at commit [`adbbc7a`](https://github.com/apache/spark/commit/adbbc7a66eabe72af2887015cbf6eb6e8553e2c6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75996/
Test PASSed.

rdblue commented:
@cloud-fan, I finally tracked down the PySpark problem. The in-memory sink had `dataframe.collect()` calls inside the streaming batch write. I think this is ready to be merged now!

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala:
```diff
@@ -281,42 +281,54 @@ class StreamExecution(
         // Unblock `awaitInitialization`
         initializationLatch.countDown()
 
-        triggerExecutor.execute(() => {
```
cc @zsxwing for streaming changes

rdblue replied:
It's hard to see from the patch, but all that's happening here is that the execution is now wrapped in `withNewExecutionID`. I also had to create an `IncrementalExecution` to pass to that method, which is like a normal one, but uses `currentBatchId=-1` (the default).

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala:
```diff
@@ -353,11 +353,19 @@ class SQLListener(conf: SparkConf) extends SparkListener with Logging {
         val driverUpdates = executionUIData.driverAccumUpdates.toSeq
-        val totalUpdates = (accumulatorUpdates ++ driverUpdates).filter {
-          case (id, _) => executionUIData.accumulatorMetrics.contains(id)
+        // filter out updates that aren't Longs
```
why?

rdblue replied:
Because all of the SQL UI metrics are longs. We can safely discard anything else without missing SQL metrics. It's as close as we can get to the correct set.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/util/DataFrameCallbackSuite.scala:
```diff
@@ -183,21 +183,22 @@ class DataFrameCallbackSuite extends QueryTest with SharedSQLContext {
+      assert(commands(2)._2.asInstanceOf[InsertIntoTable].table
         .asInstanceOf[UnresolvedRelation].tableIdentifier.table == "tab")
     }
+    // exiting withTable adds commands(3) via onSuccess (drops tab)
```
how about we update `onSuccess` to only collect commands we are interested?

rdblue replied:
I don't think it is a good idea to come up with a way to exclude commands, since this is captured by a listener. Simply noting that onSuccess will issue a drop command is sufficient.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala:
```diff
@@ -353,11 +353,19 @@ class SQLListener(conf: SparkConf) extends SparkListener with Logging {
+        // Hack alert!
+        // metrics may be missing from the executionUIData because metrics are taken from the
+        // SparkPlan passed to withNewExecutionId, but some nodes in that plan link in LogicalPlan
+        // nodes that don't expose the actual SparkPlan or metrics because the SparkPlan is built
```
for this case, can't we just ignore these metrics?

rdblue replied:
I think it's better to have the metrics than to discard them. Why discard metrics if we know that they can be safely summarized and used?

cloud-fan replied:
If these metrics are not linked with any `SparkPlan`s, will them be showed in Spark UI?

rdblue replied:
Probably not. I can revert this and ignore the test for now, since it doesn't get any metrics.

I looked at other ways around the problem, but unless we introduce a lot more changes to produce the physical plan and get its metrics, there isn't much we can do. I think it makes the most sense for those changes to to into another issue that tracks getting rid of places where the physical plan is split like this.

So what would you like me to do to get this ready to merge?

cloud-fan commented:
This PR will change the Spark UI.
For a simple query `Seq(1 -> "a").toDF("i", "j").write.parquet("/tmp/a")`, previously the SQL tab of Spark UI will show
<img width="287" alt="t1" src="https://cloud.githubusercontent.com/assets/3182036/25320581/fd74467e-28da-11e7-80ec-efb4af8a2cdb.png">
After this PR it's
<img width="577" alt="t2" src="https://cloud.githubusercontent.com/assets/3182036/25320591/116864a8-28db-11e7-9115-cf0bac552fdf.png">

I'm not sure which one is better, it depends on how users expect the Spark SQL UI for a write operation. cc @zsxwing

rdblue commented:
@cloud-fan, the one from this PR is the correct plan to show. There are two problems affecting this query. First, the call to `withNewExecutionId` should be done where the execution is started, in either `DataFrameWriter` or `Dataset`, not in the leaf nodes of the Spark plan. That way, we don't have to add calls to it everywhere and have some code paths that don't show up in the SQL tab.

The second problem is one that I discovered in tests: **queries like this have two physical plans** that are computed at different times because `ExecutedCommandExec` links a logical plan into the physical plan.

The result of both problems is that the current SQL tab only shows the second/inner physical plan, the `LocalTableScan` without showing that it is being written. The write operation is crucial information missing in the current version. The solution to this problem is to call `withNewExecutionId` at the right time (what this PR fixes) *and* to update queries like this to have a single physical plan. With that fix, this would show that the write operation read using a `LocalTableScan`.

I think the plan with the the write operation is the right plan to show, and this fixes the queries that don't show up at all. Once this goes in, I'll bring up a discussion on the dev list about how we are going to fix the queries that are split into two physical plans. That's a larger fix than this PR should handle.

zsxwing commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala:
```diff
@@ -161,50 +161,51 @@ object FileFormatWriter extends Logging {
-        committer.abortJob(job)
-        throw new SparkException("Job aborted.", cause)
+    // During tests, make sure there is an execution ID.
+    SQLExecution.checkSQLExecutionId(sparkSession)
```
The major issue is this change. For all queries using FileFormatWriter, we won't get any metrics because of https://github.com/apache/spark/blob/7536e2849df6d63587fbf16b4ecb5db06fed7125/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala#L139 . It creates a new QueryExecution and we don't track it.

zsxwing replied:
To make SQL metrics work, we should always wrap the correct QueryExecution with `SparkListenerSQLExecutionStart`.

rdblue replied:
@zsxwing, that and similar cases are what I was talking about earlier when I said there are two physical plans. The inner `Dataset.ofRows` ends up creating a completely separate plan.

Are you saying that adding `SparkListenerSQLExecutionStart` (and also end) events will fix the metrics problem? I think it would at least require the metrics work-around I added to `SQLListener`, since metrics are filtered out if they aren't reported by the physical plan.

zsxwing commented:
The current `withNewExecutionId` issue is that it doesn't support nested QueryExecution. I'm wondering if you can really fix this issue without introducing regression, e.g., track the nested QueryExecution and display them properly in the UI.

rdblue commented:
@zsxwing, what do you think the right way forward is? The "problem" with this patch seems to actually be a wide-spread problem with the data sources API -- it doesn't correctly link into SQL plans. I don't think this should be blocked by that problem since there are many queries that have no entry in the SQL tab (and also no metrics) and the current problems also cause the effects that @cloud-fan's test demonstrated where the write operation isn't shown.

I think it would be better to introduce this patch, with regressions, and follow up by fixing the data source code.

rdblue commented:
@zsxwing and @cloud-fan, can you have another look at this? I'd really like to get it in.

cloud-fan commented:
Personally, I'm fine with this patch, the only concern is we should have a follow up for nested query execution ASAP. And we should revert https://github.com/apache/spark/pull/17540#discussion_r112601926, which is just a hack for the test, as metrics without linking SparkPlan is useless, we should just fix the test instead.

rdblue committed:
SPARK-20213: Remove metrics work-around.

rdblue commented:
@cloud-fan, I've removed the metrics work-around so this is ready to go. I'm all for getting this in and fixing the broken query plans.

SparkQA commented:
**[Test build #76331 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76331/testReport)** for PR 17540 at commit [`620e697`](https://github.com/apache/spark/commit/620e6972cd31bd7a350f39ee2ac9a1d87c12d6da).

SparkQA commented:
**[Test build #76331 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76331/testReport)** for PR 17540 at commit [`620e697`](https://github.com/apache/spark/commit/620e6972cd31bd7a350f39ee2ac9a1d87c12d6da).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76331/
Test PASSed.

cloud-fan commented:
yea I think it's ready to go, @rdblue can you bring it up to date?

rdblue commented:
Rebased. I'll check if tests pass later tonight.

carsonwang commented:
Hi @rdblue , just wanted to confirm that  this also fixed #17535 so we should have the UI when executing queries in Spark SQL CLI?

carsonwang commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:
```diff
@@ -73,21 +99,35 @@ object SQLExecution {
+      // To avoid this warning, use nested { ... }
+      if (!Option(sc.getLocalProperty(ALLOW_NESTED_EXECUTION)).exists(_.toBoolean)) {
+        if (testing) {
+          logWarning(s"$EXECUTION_ID_KEY is already set")
```
According to the comment, we should throw the exception here and log warning at runtime?

rdblue replied:
Good catch. Fixed.

rdblue commented:
@carsonwang, the plan for when we notice queries that don't appear in the SQL tab is to add a call to `checkSQLExecutionId`, which will cause tests to fail when that operation isn't wrapped by `withNewExecutionId`. That should be done as a follow-up to this in #17535, since this is already getting large. Does that sound reasonable?

SparkQA commented:
**[Test build #76367 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76367/testReport)** for PR 17540 at commit [`6752ff2`](https://github.com/apache/spark/commit/6752ff292cc1958c04cbe3eaa82c4935b6b6f97f).

rdblue committed:
SPARK-20213: Fix backward logic for nesting.

SparkQA commented:
**[Test build #76368 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76368/testReport)** for PR 17540 at commit [`90045cf`](https://github.com/apache/spark/commit/90045cfe7c8ac24c1586ebe2992466bdcb2980d6).

carsonwang commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:
```diff
@@ -73,21 +99,35 @@ object SQLExecution {
+      // to ensure that nesting is avoided.
+      //
+      // To avoid this warning, use nested { ... }
+      if (!Option(sc.getLocalProperty(ALLOW_NESTED_EXECUTION)).exists(_.toBoolean)) {
```
We need check if the value is true instead of only if it exists?

carsonwang replied:
just ignore this.

carsonwang commented:
Yes, that's reasonable. I was asking because I noticed `withNewExecutionId` was added in `hiveResultString` method so it should have been fixed.

SparkQA commented:
**[Test build #76368 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76368/testReport)** for PR 17540 at commit [`90045cf`](https://github.com/apache/spark/commit/90045cfe7c8ac24c1586ebe2992466bdcb2980d6).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76368/
Test FAILed.

SparkQA commented:
**[Test build #76367 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76367/testReport)** for PR 17540 at commit [`6752ff2`](https://github.com/apache/spark/commit/6752ff292cc1958c04cbe3eaa82c4935b6b6f97f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76367/
Test PASSed.

cloud-fan commented:
The last commit fails a lot of tests...

rdblue commented:
@cloud-fan, @carsonwang pointed out that that logic for nested SQL executions was backward, so it would warn during tests and probably fail at runtime. I fixed that and now there are more test failures. I'll fix those and we should be good to go.

rdblue committed:
SPARK-20213: Fix more tests with nested SQL executions.

SparkQA commented:
**[Test build #76387 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76387/testReport)** for PR 17540 at commit [`1ce1a81`](https://github.com/apache/spark/commit/1ce1a8127caed9d886d7038b8baa327017702c22).

SparkQA commented:
**[Test build #76387 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76387/testReport)** for PR 17540 at commit [`1ce1a81`](https://github.com/apache/spark/commit/1ce1a8127caed9d886d7038b8baa327017702c22).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76387/
Test FAILed.

SparkQA commented:
**[Test build #76388 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76388/testReport)** for PR 17540 at commit [`bd324e6`](https://github.com/apache/spark/commit/bd324e604503cd08f010c19e49250c03870a67ac).

SparkQA commented:
**[Test build #76388 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76388/testReport)** for PR 17540 at commit [`bd324e6`](https://github.com/apache/spark/commit/bd324e604503cd08f010c19e49250c03870a67ac).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76388/
Test FAILed.

SparkQA commented:
**[Test build #76395 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76395/testReport)** for PR 17540 at commit [`7131c32`](https://github.com/apache/spark/commit/7131c329e34b7961ea478532faa4f202255da246).

SparkQA commented:
**[Test build #76395 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76395/testReport)** for PR 17540 at commit [`7131c32`](https://github.com/apache/spark/commit/7131c329e34b7961ea478532faa4f202255da246).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76395/
Test FAILed.

SparkQA commented:
**[Test build #76396 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76396/testReport)** for PR 17540 at commit [`69ed59e`](https://github.com/apache/spark/commit/69ed59e2d723fff9756671b390765f9106b5b720).

SparkQA commented:
**[Test build #76396 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76396/testReport)** for PR 17540 at commit [`69ed59e`](https://github.com/apache/spark/commit/69ed59e2d723fff9756671b390765f9106b5b720).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76396/
Test FAILed.

SparkQA commented:
**[Test build #76422 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76422/testReport)** for PR 17540 at commit [`4db4fc9`](https://github.com/apache/spark/commit/4db4fc9411da6a2adcd68605852116e51f22fbfa).

SparkQA commented:
**[Test build #76422 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76422/testReport)** for PR 17540 at commit [`4db4fc9`](https://github.com/apache/spark/commit/4db4fc9411da6a2adcd68605852116e51f22fbfa).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76422/
Test FAILed.

SparkQA commented:
**[Test build #76431 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76431/testReport)** for PR 17540 at commit [`f63b773`](https://github.com/apache/spark/commit/f63b773ad924fdd39eacf8ed53ece6966deba9ca).

zsxwing commented:
@rdblue I just tested this PR and found that I could not see any SQL metrics on Web UI. This is pretty important for many users to analyze their queries.

What's your plan to fix it? As far as I understand, you want to show the parameters of `write` actions, such as `InsertIntoHadoopFsRelationCommand(file://..., Parquet, ...)`. However, that's not the running `QueryExecution`. Right now we just put SQL metrics to the DAG on UI. After your change, the DAG shown on UI is not the running `QueryExecution`, and I don't know how to show SQL metrics on a wrong DAG. Is it even possible to fix it in a follow up PR quickly?

rdblue commented:
@zsxwing, there should be a fix for the metrics without waiting for all of the bad plans to be fixed (which is to basically eliminate the use of `ExecutedCommandExec`).

The metrics are missing because `ExecutedCommandExec` doesn't report them via [`metrics`](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala#L79). So we need to update it to get metrics from the command that is run. That requires breaking the command into two phases, one to get a SparkPlan and one to run it. Right now, that all happens at once. This shouldn't be too difficult as a follow-up, but will be a substantial number of changes so I think this PR should be independent.

zsxwing commented:
> That requires breaking the command into two phases, one to get a SparkPlan and one to run it.

Yeah, but how to show metrics you get from a plan on another plan's DAG considering these two plans could be different.

rdblue commented:
@zsxwing, I don't know. Sounds like we should fix the underlying problem that there are 2 physical plans.

zsxwing commented:
> @zsxwing, I don't know. Sounds like we should fix the underlying problem that there are 2 physical plans.

SQL metrics won't work without fixing it. IMO, that's more serious than the problem you are fixing.

rdblue commented:
@zsxwing, you don't think there's a way to fix metrics? I don't know exactly how to fix the UI to show two plans worth of metrics, but it seems like it can be done. What about also updating `ExecutedCommandExec` also report the plan from its child command?

Having two physical plans is a pretty bad problem for a SQL engine to have. If the work-around is to ignore that some parts of the UI don't work, I don't think that's a good plan. Sure, this is going to be a short-term regression for metrics, but what is the alternative to fix the underlying problem?

zsxwing commented:
I was not saying there is no way to fix metrics. Just asking your thoughts. If we don't have a concrete plan, it might be a long-term regression if just merging this PR.

I just want to ensure that it's going to be fixed soon after merging this PR.

rdblue commented:
I'm not an expert on the metrics path, but I think we should be able to join up the actual physical plans well enough to display everything. I doubt it will be a long-term regression, but I don't think the fix is small enough that we should include it here. I also think that it is important for this to go in so we can fix the problem. Otherwise, I think it is too easy to ignore it. Not having queries show up in the SQL tab is just as bad as the metrics issue, so I think we're trading up.

SparkQA commented:
**[Test build #76431 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76431/testReport)** for PR 17540 at commit [`f63b773`](https://github.com/apache/spark/commit/f63b773ad924fdd39eacf8ed53ece6966deba9ca).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76431/
Test PASSed.

cloud-fan commented:
how about we make `InsertIntoHadoopFsRelation` and `InsertIntoHiveTable` physical plans instead of `RunnableCommand`? Someone also asked for this refactor in https://issues.apache.org/jira/browse/SPARK-19256

rdblue commented:
@cloud-fan, all of the `RunnableCommand` instances that are currently run through `ExecutedCommandExec` need to be fixed so that there is only one physical plan. But the scope of those changes is larger than what's needed here. And as @zsxwing notes, the next step is probably to fix metrics for those broken plans.

rdblue commented:
@cloud-fan, @zsxwing, tests are passing now. Should we commit this so we can start fixing the metrics?

zsxwing commented:
I don't think we need to rush. As far as I can tell, this PR breaks two things:
- SQL Metrics on Web UI is broekn
- It doesn't display the batch queries inside a Structured Streaming query. Right now it alwayas shows one SQL query.

What benefits we can gain from this PR that are worth to break the above things?

zsxwing commented:
I suggest that you just fix them in this PR. It it has to be a large PR, I'm okey with that.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala:
```diff
@@ -283,44 +283,57 @@ class StreamExecution(
+          currentBatchId,
+          offsetSeqMetadata)
+
+        SQLExecution.withNewExecutionId(sparkSessionToRunBatches, genericStreamExecution) {
```
I think we should have an execution id for each streaming batch

cloud-fan commented:
Hi @rdblue , I have sent you a PR to fix the missing metrics issue of `ExecutedCommandExec`, and we also need to fix the streaming batch metrics issue, then we are ready to go.
